pkg/controller/logcollector/logcollector_controller.go

```diff
@@ -245,6 +247,69 @@ func (r *ReconcileLogCollector) Reconcile(request reconcile.Request) (reconcile.
 		}
 	}
 
+	if instance.Spec.AdditionalStores != nil {
+		if instance.Spec.AdditionalStores.Syslog != nil {
+			syslog := instance.Spec.AdditionalStores.Syslog
+
+			// Try to grab the ManagementClusterConnection CR because we need it for some
+			// validation with respect to Syslog.logTypes.
+			managementClusterConnection, mccErr := utils.GetManagementClusterConnection(ctx, r.client)
+			if mccErr != nil {
+				// Not finding a ManagementClusterConnection CR is not an error, as only a managed cluster will
+				// have this CR available, but we should communicate any other kind of error that we encounter.
+				if !errors.IsNotFound(err) {
+					r.status.SetDegraded(
+						"An error occurred while looking for a ManagementClusterConnection",
+						err.Error(),
+					)
+					return reconcile.Result{}, err
+				}
+			}
+
+			// If the user set Syslog.logTypes, we need to ensure that they did not include
+			// the v1.SyslogLogIDSEvents option if this is a managed cluster (i.e.
+			// ManagementClusterConnection CR is present). This is because IDS events
+			// are only forwarded within a non-managed cluster (where LogStorage is present).
+			if syslog.LogTypes != nil {
+				if mccErr == nil && managementClusterConnection != nil {
+					for _, l := range syslog.LogTypes {
+						// Set status to degraded to warn user and let them fix the issue themselves.
+						if l == v1.SyslogLogIDSEvents {
+							r.status.SetDegraded(
+								"IDSEvents option is not supported for Syslog config in a managed cluster",
+								err.Error(),
+							)
+							return reconcile.Result{}, err
+						}
+					}
+				}
+			}
+
+			// Special case: For users that have a Syslog config and are upgrading from an older release
+			//  where logTypes field did not exist, we will auto-populate default values for
+			// them. This should only happen on upgrade, since logTypes is a required field.
+			if syslog.LogTypes == nil || len(syslog.LogTypes) == 0 {
+				// Set default log types to everything except for v1.SyslogLogIDSEvents (since this
+				// option was not available prior to the logTypes field being introduced). This ensures
+				// existing users continue to get the same expected behavior for Syslog forwarding.
+				instance.Spec.AdditionalStores.Syslog.LogTypes = []v1.SyslogLogType{
+					v1.SyslogLogAudit,
+					v1.SyslogLogDNS,
+					v1.SyslogLogFlows,
+				}
+
+				// Write the default list back to the datastore.
+				if err = r.client.Update(ctx, instance); err != nil {
```

Comment:
Shouldn't this update occur outside of the "if syslog.logtypes is not specified" branch?

in most of our controllers we update the resource with the determined defaults always.

Reply:
IIUC, you're asking why have a specific update statement just for the Syslog.Logtypes, and not have a generic update statement at the end of the reconcile? Is my understanding correct? 

I could do that, if in the future there's other defaults that need to get added to LogCollector it doesn't make sense to have multiple update statements. 

Although, one advantage to having a specific update here is that set degraded message can be specific and thus potentially more helpful to the user.

Reply:
> IIUC, you're asking why have a specific update statement just for the Syslog.Logtypes, and not have a generic update statement at the end of the reconcile? Is my understanding correct?

Yep, in general the pattern we have established for controllers in the operator is that it:

- Queries the CR
- Populates any defaults
- Performs validation
- Updates the CR in the API
- Renders resources

Really well exemplified in the installation controller: https://github.com/tigera/operator/blob/master/pkg/controller/installation/core_controller.go#L581-L600

We probably should have been doing this in this controller from the get-go.

Basically, the reason for always writing it back after defaults is so that we can ensure consistent behavior for users on upgrade, even if we change what the code itself is doing. If a users has LogTypes=[X,Y,Z] in version A, and we change the defaults in version B, the user can upgrade from A->B without fear of behavior changing unexpectedly, since their desired state is spelled out explicitly in the config. 

I haven't done a comprehensive review of this controller in its current form, so maybe I am way off-base, but if there are any other values that we're defaulting in the "render" stage we should be moving those up to the "populate defaults" stage so we can write them back (not in this PR obviously, but just in general).

Anyway, a long-winded response to basically say "Even though we're only doing it for one field right now, we should be doing it for all of them so lets structure the code with that in mind"

Reply:
Cool, thanks for the explainer with the example. I'll be sure to remember the general pattern you outlined with the example moving forward. 

Makes sense to me. I've moved the update out so it's generic.

Reply:
It looks like you've put the "Updates the CR in the API" after the "Renders resources". I think the other controllers do the "Updates the CR in the API" as soon as the CR has been updated, so before even retrieving and validating that the config like Secrets and other CRs are valid.